delaymodel.py

```diff
@@ -229,64 +231,83 @@ def calculate_delay(self, publish : bool = True):
+                self.source = SkyCoord(self.source_pointings_dict['ra'], self.source_pointings_dict['dec'], unit='deg')
+                loadtime = self.source_pointings_dict['loadtime']
+                sideband = self.source_pointings_dict['sideband']
+                sslo = self.source_pointings_dict['sslo']
+
+            elif loadtime_from_now is None:
+                #loadtime is None. Update source pointing and calculate coefficients and set loadtime to 
+                #self.min_loadtime_offset from now.
+                self.source = SkyCoord(self.source_pointings_dict['ra'], self.source_pointings_dict['dec'], unit='deg')
+                loadtime = t_int + self.min_loadtime_offset
+                sideband = self.source_pointings_dict['sideband']
+                sslo = self.source_pointings_dict['sslo']
+            
+            else:
+                logger.error("Invalid loadtime provided")
+
+            tts = [3, (TIME_INTERPOLATION_LENGTH/2) + 3, TIME_INTERPOLATION_LENGTH + 3]
+            tts = np.array(tts) + t_int # Interpolate time samples with 3s advance
+            dt = TIME_INTERPOLATION_LENGTH/2
+            ts = Time(tts, format='unix')
+
+            # perform coordinate transformation to uvw
+            uvw1 = compute_uvw(ts[0], self.source, self.itrf[['X','Y','Z']], ITRF_CENTER)
+            uvw2 = compute_uvw(ts[1], self.source, self.itrf[['X','Y','Z']], ITRF_CENTER)
+            uvw3 = compute_uvw(ts[2], self.source, self.itrf[['X','Y','Z']], ITRF_CENTER)
+
+            # "w" coordinate represents the goemetric delay in light-meters
+            w1 = uvw1[...,2]
+            w2 = uvw2[...,2]
+            w3 = uvw3[...,2]
+
+            # Calibration delays are added in the controller
+            delay1 = (w1/const.c.value)
+            delay2 = (w2/const.c.value)
+            delay3 = (w3/const.c.value)
+
+            # Compute the delay rate in s/s
+            rate1 = (delay2 - delay1) / (dt)
+            rate2 = (delay3 - delay2) / (dt)
+            rate = (delay3 - delay1) / (2*dt)
+
+            # Compute the delay rate rate in s/s^2
+            raterate = (rate2 - rate1) / (dt)
+
+            self.delay_data["delay_ns"] = (delay2*1e9).tolist()
+            self.delay_data["delay_rate_nsps"] = (rate*1e9).tolist()
+            self.delay_data["delay_raterate_nsps2"] = (raterate*1e9).tolist()
+            self.delay_data["effective_lo_0_mhz"] = sslo[0]
+            self.delay_data["effective_lo_1_mhz"] = sslo[1]
+            self.delay_data["sideband_0"] = sideband[0]
+            self.delay_data["sideband_1"] = sideband[1]
+            self.delay_data["time_value"] = tts[1]
+            self.delay_data["loadtime"] = loadtime
+
+            if publish:
+                self.publish_delays()
+            else:
+                return pd.DataFrame(self.delay_data, index=list(self.itrf.index.values)).to_dict('index')               
+
+            time.sleep(time.time() + self.max_loadtime_offset - t)
```

Comment:
Are you sure this doesn't block any updates? I.e. if the code is here, sleeping for 10 seconds, and 1 second in a user submits a request with a load time 5 seconds in the future, what happens?

Reply:
I think the laziest pattern (which amounts to doing lots of unnecessary compute work) is:
```
last_publish_time = 0
while True:
  # figure out delays
  if publish:
    # self.new is set whenever a new pointing directive is received from redis
    if self.new or (time.time() - last_publish_time) > self.max_loadtime_offset: # should really be it's own "publish period" variable, probably
      self.publish_delays()
  time.sleep(1) # Needs to be less than the minimum possible loadtime
```

Reply:
Good point, that `sleep` is blocking. So, while the message will be received (the listener runs in a thread), the message won't be addressed until after the sleep. 

Dropping the `sleep` duration however could mean that we re-interpolate too shortly after just pushing out coefficients if the pointing does **not** change...

So perhaps I need to sleep for `1s` at the end of the loop and then in [this](https://github.com/COSMIC-SETI/COSMIC-VLA-DelayEngine/blob/delaymodelupdate/delaymodel.py#L239) if statement, sleep for an additional period because that is the case for when we have not received new pointings... this sleep would however also be blocking :/